source/extensions/filters/http/aws_request_signing/aws_request_signing_filter.cc

```diff
@@ -43,6 +55,35 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   return Http::FilterHeadersStatus::Continue;
 }
 
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+  if (config_->useUnsignedPayload()) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (!end_stream) {
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+
+  decoder_callbacks_->addDecodedData(data, false);
+
+  const Buffer::Instance& decoding_buffer = *decoder_callbacks_->decodingBuffer();
+
+  auto& hashing_util = Envoy::Common::Crypto::UtilitySingleton::get();
+  const std::string hash = Hex::encode(hashing_util.getSha256Digest(decoding_buffer));
+
+  try {
+    ENVOY_LOG(debug, "aws request signing from decodeData");
+    ASSERT(request_headers_ != nullptr);
+    config_->signer().sign(*request_headers_, hash);
+    config_->stats().signing_added_.inc();
```

Comment:
Hmm should we add a new stat `signing_body_added` or similar? To distinguish from the existing signing code path...

Reply:
Having separate counters could add fidelity when `unsigned_payload` is `false`. I don't actually know under what circumstances `end_stream` is `true` in `decodeHeaders`, so I suppose having discrete counters could be useful in diagnosing any unexpected behavior (like if there's some way `end_stream` is `true` when there was a request body).